drivers/i2c/i2c_ll_stm32_v2.c

```diff
@@ -791,6 +825,185 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+	cr2 &= ~(I2C_CR2_RELOAD | I2C_CR2_AUTOEND | I2C_CR2_NBYTES_Msk | I2C_CR2_SADD_Msk);
+
+	if (I2C_ADDR_10_BITS & data->dev_config) {
+		cr2 |= (uint32_t)slave | I2C_CR2_ADD10;
+	} else {
+		cr2 |= (uint32_t)slave << 1;
+	}
+
+	/*
+	 * If this is not a stop message and more messages follow without change of direction,
+	 * reload mode must be used during this transaction
+	 * also a helper variable is set to inform IRQ handler about that it should
+	 * keep reload mode turned on ready for next message
+	 */
+	if (!(msg->flags & I2C_MSG_STOP) && next_msg_flags &&
+	    !(*next_msg_flags & I2C_MSG_RESTART)) {
+		cr2 |= I2C_CR2_RELOAD;
+		data->current.continue_in_next = true;
+	} else {
+		data->current.continue_in_next = false;
+	}
+
+	/*
+	 * For messages larger than 255 bytes, transactions must be split i chunks
+	 * Use reload mode and let IRQ handler take care of jumping to next chunk
+	 */
+	if (msg->len > 255U) {
+		cr2 |= (255UL << I2C_CR2_NBYTES_Pos) | I2C_CR2_RELOAD;
+	} else {
+		/* Whole message can be sent in one I2C HW transaction */
+		cr2 |= msg->len << I2C_CR2_NBYTES_Pos;
+	}
+
+	/*
+	 * If a reload mode transfer is pending since last message then skip
+	 * checking for transfer complete or restart flag in message
+	 * Reload transfer will start right after writing new length
+	 * to CR2 below
+	 */
+	if ((isr & I2C_ISR_TCR) == 0) {
+
+		/*
+		 * As TCR is not set, expect TC to be set or that this is a (re)start message
+		 * - msg->flags contains I2C_MSG_RESTART (for first start) or
+		 * - TC in ISR register is set which happens when IRQ handler
+		 *   has finalized its transfer and is waiting for restart
+		 * For both cases, a new start condition shall be sent
+		 */
+		__ASSERT_NO_MSG(((isr & I2C_ISR_TC) != 0U) || (msg->flags & I2C_MSG_RESTART));
+
+		if ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+			cr2 &= ~I2C_CR2_RD_WRN;
+			/*
+			 * Prepare first byte in TX buffer before transfer start as a
+			 * workaround for errata: "Transmission stalled after first byte transfer"
+			 */
+			LL_I2C_TransmitData8(regs, *data->current.buf);
+			data->current.len--;
+			data->current.buf++;
+
+		} else {
+			cr2 |= I2C_CR2_RD_WRN;
+		}
+		/* Issue (re)start condition */
+		cr2 |= I2C_CR2_START;
+	}
+
+	/* Commit to I2C controller */
+	LL_I2C_WriteReg(regs, CR2, cr2);
+
+	uint32_t cr1 = LL_I2C_ReadReg(regs, CR1);
+	/* Set common interrupt enable bits */
+	cr1 |= I2C_CR1_ERRIE | I2C_CR1_STOPIE | I2C_CR1_TCIE | I2C_CR1_NACKIE;
+
+#ifdef CONFIG_I2C_STM32_V2_DMA
+	dma_xfer_start(dev, msg);
+#else
+	/* If not using DMA, also enable RX and TX empty interrupts */
+	cr1 |= I2C_CR1_TXIE | I2C_CR1_RXIE;
+#endif
+	/* Enable interrupts */
+	LL_I2C_WriteReg(regs, CR1, cr1);
+
+	/* Wait for IRQ to complete or timeout
+	 * Timeout scales with one millisecond for each byte to
+	 * transfer so that slave can do some clock stretching
+	 */
+	if (k_sem_take(&data->device_sync_sem, K_MSEC(msg->len + 10)) != 0) {
+		is_timeout = true;
+	}
```

Comment:
Call to `dma_stop()` is missing here. Despite it's brought back in a next commit through `dma_finish()`, the current commit should support DMA (to prevent later issue, e.g. when using `git bisect`).